src/api/mod.rs

```diff
@@ -403,6 +438,168 @@ impl<'a> ValueListBuilder<'a> {
     }
 }
 
+fn to_meta_data<'a, T>(meta_hm: T) -> Result<*mut meta_data_t, SubmitError>
+where
+    T: IntoIterator<Item = (&'a &'a str, &'a MetaValue)>,
+{
+    let meta = unsafe { meta_data_create() };
+    let conversion_result = to_meta_data_with_meta(meta_hm, meta);
+    match conversion_result {
+        Ok(()) => Ok(meta),
+        Err(error) => {
+            unsafe {
+                meta_data_destroy(meta);
+            }
+            Err(error)
+        }
+    }
+}
+
+fn to_meta_data_with_meta<'a, T>(meta_hm: T, meta: *mut meta_data_t) -> Result<(), SubmitError>
+where
+    T: IntoIterator<Item = (&'a &'a str, &'a MetaValue)>,
+{
+    for (key, value) in meta_hm.into_iter() {
+        let c_key = CString::new(*key).map_err(|e| {
+            SubmitError::Field(
+                "meta key",
+                ArrayError::NullPresent(e.nul_position(), key.to_string()),
+            )
+        })?;
+        match value {
+            MetaValue::String(str) => {
+                let c_value = CString::new(str.as_str()).map_err(|e| {
+                    SubmitError::Field(
+                        "meta value",
+                        ArrayError::NullPresent(e.nul_position(), str.to_string()),
+                    )
+                })?;
+                unsafe {
+                    meta_data_add_string(meta, c_key.as_ptr(), c_value.as_ptr());
+                }
+            }
+            MetaValue::SignedInt(i) => unsafe {
+                meta_data_add_signed_int(meta, c_key.as_ptr(), *i);
+            },
+            MetaValue::UnsignedInt(u) => unsafe {
+                meta_data_add_unsigned_int(meta, c_key.as_ptr(), *u);
+            },
+            MetaValue::Double(d) => unsafe {
+                meta_data_add_double(meta, c_key.as_ptr(), *d);
+            },
+            MetaValue::Boolean(b) => unsafe {
+                meta_data_add_boolean(meta, c_key.as_ptr(), *b);
+            },
+        }
+    }
+    Ok(())
+}
+
+fn from_meta_data(
+    p: &str,
+    meta: *mut meta_data_t,
+) -> Result<HashMap<String, MetaValue>, ReceiveError> {
+    if meta.is_null() {
+        return Ok(HashMap::new());
+    }
+
+    let mut c_toc: *mut *mut c_char = ptr::null_mut();
+    let count_or_err = unsafe { meta_data_toc(meta, &mut c_toc as *mut *mut *mut c_char) };
+    if count_or_err < 0 {
+        return Err(ReceiveError::Metadata(
+            p.to_string(),
+            "toc".to_string(),
+            "invalid parameters to meta_data_toc",
+        ));
+    }
+    let count = count_or_err as usize;
+    if count == 0 {
+        return Ok(HashMap::new());
+    }
+
+    let toc = unsafe { slice::from_raw_parts(c_toc, count) };
+    let conversion_result = from_meta_data_with_toc(p, meta, toc);
+
+    for c_key_ptr in toc {
```

Comment:
I was going to say that this loop and the subsequent free can be replaced with a single call to `strarray_free` as seen here: https://github.com/collectd/collectd/blob/1da3305c10c8ff9a63081284cf3d4bb0f6daffd8/src/grpc.cc#L233 However it looks like `strarray_free` is not distributed in the `collectd-dev` package, I'm reluctant to depend on it.

A convention that can be used from that linked example is that even when there are zero entries, the array is still freed so if you wanted to remove the `if count == 0` special case, you certainly can and it would still match collectd conventions. Up to you, I have no preference.

Reply:
i'll leave it as it is, because if `count == 0` meta_data_toc does not allocate anything and i should guard the `free(c_toc)` line against null pointer as collectd does with `sfree()`